test/Tests/src/Semantic/Case_Spec.enso

```diff
@@ -163,4 +215,67 @@ spec = Test.group "Pattern Matches" <|
             Cons _ Nil            -> Test.fail "Expected value to match constant."
             _                     -> Test.fail "Expected value to match constant."
 
+    Test.specify "should be able to match on module rather than a type" <|
+        case Vector of
+#            _ : Vector       -> Test.fail "Expected to match on module."
+            _ : Vector.Vector -> Test.fail "Expected to match on module."
+            Vector.Vector     -> Test.fail "Expected to match on module."
+            Vector            -> Nothing
+            _                 -> Test.fail "Expected to match on module."
+
+    Test.specify "should be able to match on a type of a value rather tha module" <|
+        case [1,2,3] of
+#            _ : Vector     -> Test.fail "Expected to match on a type."
+            Vector          -> Test.fail "Expected to match on a type."
+            Vector.Vector   -> Test.fail "Expected to match on a type."
+            _ : Vector.Vector -> Nothing
+            _               -> Test.fail "Expected to match on a type."
+
+    Test.specify "should correctly pattern match on array-like polyglot values" <|
+        list = ArrayList.new
+        list.add 1
+        list.add 2
+
+        case list of
+            _ : Vector.Vector -> Test.fail "Expected to match on Array type."
+            _ : Array -> Nothing
+            _ : ArrayList -> Test.fail "Expected to match on Array type."
+            _ -> Test.fail "Expected to match on Array type."
+
+        case list of
+            _ : Vector.Vector -> Test.fail "Expected to match on ArrayList type."
+            _ : ArrayList -> Nothing
```

Comment:
Moreover `java.util.ArrayList` is also of `java.util.ArrayList` type. Great! This is going to be useful when dealing with exceptions - catching Enso as well as polyglot exception shall now be possible with the same `case_of` construct, I hope. CCing @kustosz and @radeusgd.